File: queues/sqs/Node/sqsclient.js
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Amazon SQS helpers for Core Auto. Send from step scripts; Receive for ingress bridges.

import {
  SQSClient,
  SendMessageCommand,
  ReceiveMessageCommand,
  DeleteMessageCommand,
} from '@aws-sdk/client-sqs';
import { missingEnv, transportError } from './lib/result.js';

function client() {
  const cfg = { region: process.env.AWS_REGION || process.env.AWS_DEFAULT_REGION || 'us-east-1' };
  if (process.env.SQS_ENDPOINT_URL) cfg.endpoint = process.env.SQS_ENDPOINT_URL;
  return new SQSClient(cfg);
}

function queueUrl(explicit) {
  return explicit || process.env.SQS_QUEUE_URL || '';
}

function encode(v) {
  return typeof v === 'string' ? v : JSON.stringify(v);
}

function decode(raw) {
  try {
    return JSON.parse(raw);
  } catch {
    return raw;
  }
}

export function Init() {
  if (!process.env.AWS_ACCESS_KEY_ID && !process.env.AWS_PROFILE) {
    return missingEnv('AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_PROFILE');
  }
  if (!process.env.SQS_QUEUE_URL) {
    return missingEnv('SQS_QUEUE_URL (or pass queue_url per call)');
  }
  return { status_code: 200 };
}

export async function Send(value, queueUrlArg = null) {
  const url = queueUrl(queueUrlArg);
  if (!url) return missingEnv('SQS_QUEUE_URL');
  try {
    const resp = await client().send(
      new SendMessageCommand({ QueueUrl: url, MessageBody: encode(value) }),
    );
    return { status_code: 200, message_id: resp.MessageId };
  } catch (exc) {
    return transportError(String(exc.message || exc));
  }
}

export async function Receive(queueUrlArg = null, maxMessages = 1, waitTimeSec = 10, deleteMsg = true) {
  const url = queueUrl(queueUrlArg);
  if (!url) return missingEnv('SQS_QUEUE_URL');
  try {
    const resp = await client().send(
      new ReceiveMessageCommand({
        QueueUrl: url,
        MaxNumberOfMessages: Math.min(Math.max(maxMessages, 1), 10),
        WaitTimeSeconds: waitTimeSec,
      }),
    );
    const messages = [];
    for (const item of resp.Messages || []) {
      messages.push({
        message_id: item.MessageId,
        receipt_handle: item.ReceiptHandle,
        value: decode(item.Body || ''),
      });
      if (deleteMsg && item.ReceiptHandle) {
        await client().send(
          new DeleteMessageCommand({ QueueUrl: url, ReceiptHandle: item.ReceiptHandle }),
        );
      }
    }
    return { status_code: 200, messages };
  } catch (exc) {
    return transportError(String(exc.message || exc));
  }
}

File: queues/sqs/Python/sqsclient.py
"""
Copyright Core DF

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Amazon SQS helpers for Core Auto. **Send** from step scripts; **Receive** for
[ingress](../../ingress/Python/README.md) bridges only (not steps).
"""

import json
import os
from typing import Any, Optional

from lib.result import missing_env, transport_error


def _client():
    import boto3

    region = os.environ.get("AWS_REGION", os.environ.get("AWS_DEFAULT_REGION", "us-east-1"))
    endpoint = os.environ.get("SQS_ENDPOINT_URL", None)
    kwargs = {"region_name": region}
    if endpoint:
        kwargs["endpoint_url"] = endpoint
    return boto3.client("sqs", **kwargs)


def _queue_url(explicit: Optional[str]) -> str:
    return explicit or os.environ.get("SQS_QUEUE_URL", "")


def _encode(value: Any) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, (dict, list)):
        return json.dumps(value)
    return str(value)


def _decode(raw: str) -> Any:
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return raw


def Init() -> dict:
    if not os.environ.get("AWS_ACCESS_KEY_ID") and not os.environ.get("AWS_PROFILE"):
        return missing_env("AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_PROFILE")
    if not os.environ.get("SQS_QUEUE_URL"):
        return missing_env("SQS_QUEUE_URL (or pass queue_url per call)")
    return {"status_code": 200}


def Send(value: Any, queue_url: Optional[str] = None) -> dict:
    url = _queue_url(queue_url)
    if not url:
        return missing_env("SQS_QUEUE_URL")
    try:
        client = _client()
        resp = client.send_message(QueueUrl=url, MessageBody=_encode(value))
        return {"status_code": 200, "message_id": resp.get("MessageId")}
    except Exception as exc:
        return transport_error(str(exc))


def Receive(
    queue_url: Optional[str] = None,
    max_messages: int = 1,
    wait_time_sec: int = 10,
    delete: bool = True,
) -> dict:
    """Long-poll messages from a queue. For ingress processes only — not step scripts."""
    url = _queue_url(queue_url)
    if not url:
        return missing_env("SQS_QUEUE_URL")
    max_messages = max(1, min(max_messages, 10))
    try:
        client = _client()
        resp = client.receive_message(
            QueueUrl=url,
            MaxNumberOfMessages=max_messages,
            WaitTimeSeconds=wait_time_sec,
        )
        messages = []
        for item in resp.get("Messages", []):
            messages.append(
                {
                    "message_id": item.get("MessageId"),
                    "receipt_handle": item.get("ReceiptHandle"),
                    "value": _decode(item.get("Body", "")),
                }
            )
            if delete and item.get("ReceiptHandle"):
                client.delete_message(QueueUrl=url, ReceiptHandle=item["ReceiptHandle"])
        return {"status_code": 200, "messages": messages}
    except Exception as exc:
        return transport_error(str(exc))
